data_safe_haven/external/api/graph_api.py

```diff
@@ -573,12 +593,145 @@ def get_id_from_username(self, username: str) -> str | None:
                 next(
                     user
                     for user in self.read_users()
-                    if user["mailNickname"] == username
+                    if user["userPrincipalName"].split("@")[0] == username
                 )["id"]
             )
         except (DataSafeHavenMicrosoftGraphError, StopIteration):
             return None
 
+    def grant_role_permissions(
+        self,
+        application_name: str,
+        *,
+        application_role_assignments: Sequence[str],
+        delegated_role_assignments: Sequence[str],
+    ) -> None:
+        """Grant admin permission for requested application roles"""
+        # Ensure that the application has a service principal
+        self.ensure_application_service_principal(application_name)
+
+        # Grant any requested application role permissions
+        for role_name in application_role_assignments:
+            self.grant_application_role_permissions(application_name, role_name)
+
+        # Grant any requested delegated role permissions
+        for role_name in delegated_role_assignments:
+            self.grant_delegated_role_permissions(application_name, role_name)
+
+    def grant_application_role_permissions(
+        self, application_name: str, application_role_name: str
+    ) -> None:
+        """
+        Grant permissions for a particular role to an application.
+        See https://learn.microsoft.com/en-us/graph/permissions-grant-via-msgraph
+
+        Raises:
+            DataSafeHavenMicrosoftGraphError if the secret could not be created or already exists
+        """
+        try:
+            # Get service principals for Microsoft Graph and this application
+            microsoft_graph_sp = self.get_service_principal_by_name("Microsoft Graph")
+            if not microsoft_graph_sp:
+                msg = "Could not find Microsoft Graph service principal."
+                raise DataSafeHavenMicrosoftGraphError(msg)
```

Comment:
It probably means that Microsoft have done a major behind-the-scenes refactor, since the Microsoft Graph application (and service principal) are installed by default in all EntraID directories and are needed to make some REST calls (like this one).

Do you have a suggested better wording?

Reply:
What I'm thinking of is something to make sure the user knows it isn't a configuration problem. Like how `pip` will say something like "this is a problem with the package you are trying to install, non pip".

Maybe "This is likely a problem with dsh, not your configuration or infrastructure"?

Reply:
Done